control/mateqn.py

```diff
@@ -623,191 +557,150 @@ def dare(A, B, Q, R, S=None, E=None, stabilizing=True):
     state space operations.  See :func:`~control.use_numpy_matrix`.
 
     """
-    if S is not None or E is not None or not stabilizing:
-        return dare_old(A, B, Q, R, S, E, stabilizing)
+    # Decide what method to use
+    method = _slycot_or_scipy(method)
+
+    # Reshape input arrays
+    A = np.array(A, ndmin=2)
+    B = np.array(B, ndmin=2)
+    Q = np.array(Q, ndmin=2)
+    R = np.eye(B.shape[1]) if R is None else np.array(R, ndmin=2)
+    if S is not None:
+        S = np.array(S, ndmin=2)
+    if E is not None:
+        E = np.array(E, ndmin=2)
+
+    # Determine main dimensions
+    n = A.shape[0]
+    m = B.shape[1]
+
+    # Check to make sure input matrices are the right shape and type
+    _check_shape("A", A, n, n, square=True)
+
+    # Figure out how to solve the problem
+    if method == 'scipy' and not stabilizing:
+        raise ValueError(
+            "method='scipy' not valid when stabilizing is not True")
+
+    elif method == 'slycot':
+        return _dare_slycot(A, B, Q, R, S, E, stabilizing)
+
     else:
+        _check_shape("B", B, n, m)
+        _check_shape("Q", Q, n, n, square=True, symmetric=True)
```

Comment:
Maybe `_dare_slycot()` and `_dare_scipy()` should be on the same level?

Reply:
I agree the asymmetry is annoying.  I left it this way to avoid changing code that was working.  In `care` everything is in the main function, so at some point we should clean all of these up to be more uniform (either a single function or use `_method` uniformly).